### PR TITLE
boot_with_different_vectors.py: add windows support

### PR DESCRIPTION
**Pull request by kimi1978wy**
Add windows support to boot with diifferent vectors

Signed-off-by: Yu Wang <wyu@redhat.com>

id:1485220

### CODE REVIEW

File: qemu/tests/cfg/boot_with_different_vectors.cfg [hunk 1]
```diff
@@ -19,5 +15,27 @@
...
+        no Host_RHEL.m5, Host_RHEL.m6
+        msi_check_cmd = "lspci -vvv -s %s | grep MSI"
+        irq_check_cmd = "cat /proc/interrupts | grep virtio"
+
+    Windows:
+        no 2k8
+        cdroms += " virtio"
+        traceview_version = "win8"
+        Win10:
+            traceview_version = "win10"
```
Comment (xiagao) [inline qemu/tests/cfg/boot_with_different_vectors.cfg]: Only Win10 guest use this version?

File: qemu/tests/cfg/boot_with_different_vectors.cfg [hunk 2]
```diff
@@ -19,5 +15,27 @@
...
+
+    Linux:
+        no RHEL.3 RHEL.4 RHEL.5 RHEL.6
+        no RHEL.7.0 RHEL.7.1 RHEL.7.2 RHEL.7.3 RHEL.7.4
+        no Host_RHEL.m5, Host_RHEL.m6
+        msi_check_cmd = "lspci -vvv -s %s | grep MSI"
+        irq_check_cmd = "cat /proc/interrupts | grep virtio"
+
+    Windows:
+        no 2k8
```
Comment (xiagao) [inline qemu/tests/cfg/boot_with_different_vectors.cfg]: Does it mean not support win2008-32,win2008-64 and win2008r2 ?

File: qemu/tests/boot_with_different_vectors.py [hunk 3]
```diff
@@ -45,41 +49,168 @@ def boot_guest_with_vectors(vectors):
...
+            # use traceview
+            prepare_traceview_windows(session, 360)
+            msis, queues = get_MSIs_and_queues_windows()
+            if msis is None or queues is None:
+                test.fail("Can't get msi status from guest.")
+            if vectors == 0 and msis != 0 and queues != 1:
+                test.fail("Msis should be 0, queues should be 1  when vectors = 1,"
+                          " but guest msis = %s, queues = %s" % (msis, queues))
+            elif vectors != msis:
+                test.fail("Msis should equal to vectors(%s), but guest is %s" % (vectors, msis))
```
Comment (xiagao) [inline qemu/tests/boot_with_different_vectors.py]: Could you add else conditon

File: qemu/tests/boot_with_different_vectors.py [hunk 4]
```diff
@@ -45,41 +49,168 @@ def boot_guest_with_vectors(vectors):
...
+        # verify driver
+        error_context.context("Check if the driver is installed and "
+                              "verified", logging.info)
+        driver_name = params.get("driver_name", "netkvm")
+        session = utils_test.qemu.windrv_check_running_verifier(session, vm,
+                                                                test,
+                                                                driver_name,
+                                                                timeout)
+
+        # copy traceview.exe
```
Comment (xiagao) [inline qemu/tests/boot_with_different_vectors.py]: also can use the absolute path but not copy to c:\，right?

File: qemu/tests/boot_with_different_vectors.py [hunk 5]
```diff
@@ -45,41 +49,168 @@ def boot_guest_with_vectors(vectors):
         vm.verify_alive()
         return vm
 
+    def get_MSI_queue_from_traceview_output(output):
+
```
Comment (xiagao) [inline qemu/tests/boot_with_different_vectors.py]: Could you add docstring for these functions, which will be more clear.thx

Comment (kimi1978wy) [reply qemu/tests/boot_with_different_vectors.py]: Thanks, modified

File: qemu/tests/boot_with_different_vectors.py (hunk 4 above)
Comment (kimi1978wy) [reply qemu/tests/boot_with_different_vectors.py]: Yes, but I think copy to c:\ is more stable.

File: qemu/tests/boot_with_different_vectors.py (hunk 3 above)
Comment (kimi1978wy) [reply qemu/tests/boot_with_different_vectors.py]: But the "else " is only pass this case. So I think it no need to add this condition

**Comment by kimi1978wy** [comment]
@xiagao @xiywang  could you help to review this patch?

File: qemu/tests/boot_with_different_vectors.py [hunk 6]
```diff
@@ -45,41 +49,192 @@ def boot_guest_with_vectors(vectors):
...
+                            logging.info("MSI-X is enabled")
+                        else:
+                            msg = "Command %s get wrong output." % msi_check_cmd
+                            msg += " when vectors = %d" % vectors
+                            test.fail(msg)
+        else:
+            # use traceview
+            prepare_traceview_windows(session, 360)
+            msis, queues = get_MSIs_and_queues_windows()
+            if msis is None or queues is None:
```
Comment (hereischen) [inline qemu/tests/boot_with_different_vectors.py]: if None in (msis,queues )

File: qemu/tests/boot_with_different_vectors.py [hunk 7]
```diff
@@ -45,41 +49,192 @@ def boot_guest_with_vectors(vectors):
...
+        to trigger logging. By analyzing the dumped output, the msi& queue
+        info is acquired.
+
+        :return: a tuple of (msis, queues)
+        """
+
+        session_serial = vm.wait_for_serial_login(login_timeout)
+        # start traceview
+        error_context.context("Start trace view with pdb file", logging.info)
+        log_path = "c:\\logfile.etl"
```
Comment (hereischen) [inline qemu/tests/boot_with_different_vectors.py]: Some parts of the code can be reused in this function. Eg:
```python
def _remove_file(file_path):
         clean_cmd = "del %s" % file_path
         session_serial.cmd(clean_cmd)
```
You can define some small function like above to make the code easier to read, and reduce string formats. Some actions are symmetry, functions would help a little bit.

But the current implantation is OK to me.

**Comment by kimi1978wy** [comment]
(1/1) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win8.i386.1.io-github-autotest-qemu.boot_with_different_vectors.q35: PASS (2203.11 s)

File: qemu/tests/boot_with_different_vectors.py [hunk 8]
```diff
@@ -45,41 +49,214 @@ def boot_guest_with_vectors(vectors):
         vm.verify_alive()
         return vm
 
+    def get_MSI_queue_from_traceview_output(output):
+        """
+        Extract msi&queues infomation from traceview log file output
```
Comment (xiywang) [inline qemu/tests/boot_with_different_vectors.py]: typo information

**Comment by hereischen** [comment]
Acked-by: Haotong Chen hachen@redhat.com

**Pull request merged**